Add basic openQA test file for SUMA containers

m4u9 opened the pull request:
Implement basic openQA testing for SUMA container images using podman, based on the rancher container test.

 The test  covers:
- Pull the image from internal  registry.suse.de test registry
- Executes a simple command within the container to test that it at least runs

The test is meant for  the following SUMA container images:

- proxy-httpd-image
- proxy-salt-broker-image
- proxy-squid-image
- proxy-ssh-image
- proxy-tftpd-image
- proxy-helm charts

Additional info:
- Based on work done on Draft PR #17425
- Related ticket: https://progress.opensuse.org/issues/132692

ricardobranco777 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,34 @@
+
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    assert_script_run("podman pull $image");
```
Depending on the size of the image you may want to add something like this (copied from [here](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/containers/podman_network_cni.pm)

```suggestion
    assert_script_run("podman pull $image", timeout => 300, delay => 60, retry => 3););
```

pdostal replied:
@ricardobranco777 you ment?
```suggestion
    script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
```

ricardobranco777 replied:
> @ricardobranco777 you ment?

Yes.  Thanks for pointing it out.

m4u9 replied:
One more thing, what does `script_retry` do exactly or where can I find the various definitions?

ricardobranco777 replied:
> One more thing, what does `script_retry` do exactly or where can I find the various definitions? I found this http://open.qa/api/testapi/ however script_retry is missing

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L1846

ricardobranco777 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,34 @@
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    assert_script_run("podman pull $image");
+    assert_script_run("podman run $image cat /etc/os-release");
```
```suggestion
    assert_script_run("podman run --rm $image cat /etc/os-release");
```

To remove the container after running it.

m4u9 replied:
Thanks, I'll add this :)

ricardobranco777 reviewed (approved):
LGTM.  Beware that if you're not using a pull-through cache registry you may run into Docker Hub rate limits.

ricardobranco777 commented:
> I have updated the commit with all the recommendations

CI is failing. Please check.

jlausuch commented on lib/main_containers.pm:
```diff
@@ -218,6 +218,11 @@ sub load_container_tests {
         return;
     }
 
+    if (get_var(' CONTAINER_SUMA')) {
```
```suggestion
    if (get_var('CONTAINER_SUMA')) {
```

jlausuch commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+      script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
```
```suggestion
    script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
```
You need to fix the formatting. 4 spaces

m4u9 commented:
> > I have updated the commit with all the recommendations
> 
> CI is failing. Please check.

Applied some changes i.e fixed indentation and removed trailing white space

jlausuch commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
```
```suggestion
    if (check_var('CONTAINER_SUMA', 'image')) {
        script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
        assert_script_run("podman run --rm $image cat /etc/os-release");
    }
```

jlausuch commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+    script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
+    assert_script_run("podman run --rm $image cat /etc/os-release");
+    }
+    elsif (check_var('CONTAINER_SUMA', 'chart')) {
```
```suggestion
    elsif (check_var('CONTAINER_SUMA', 'chart')) {
        script_retry("helm pull $image", timeout => 300, delay => 60, retry => 3);
        assert_script_run("tar -tf proxy-*.tgz && rm proxy-*.tgz");
    }
```

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+}
+
+sub test_flags {
+    return {fatal => 1};
```
Do we really want to mark this test module as fatal, i.e. the whole test run will be stopped and other tests won't be executed?

m4u9 replied:
The tests check that the container image exists and runs. I think it's ok to stop here if  either of these two basic tests fail. Or did I not understand your remark?

grisu48 replied:
`fatal => 1` stops the whole test run. So if this is expected to be a short test run that consists only of this test module, this is IMHO not needed. If this would be in a larger test run, where e.g. you would execute this next to other container tests, then all of the following ones would be skipped.

So from my POV - in the former case it would not matter, in the latter case it would be undesired. So I'd argue, unless necessary, this should not be there.

WDYT?

m4u9 replied:
I included it because it was in the Rancher test I based this from. If it's not needed I'm ok to remove it, I don't know know openQA well enough to have a strong opinion on this :-)

m4u9 replied:
I removed the fatal => 1 part

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
+        assert_script_run("podman run --rm $image cat /etc/os-release");
```
Improvement suggestion: Use [validate_script_output](https://open.qa/api/testapi/#_validate_script_output) instead of `assert_script_run` and check, that `/etc/os-release` contains what it's supposed to be containing.

That's a suggestion, feel free to ignore it.

m4u9 replied:
The purpose of the test is only to ensure the container is running and we can run commands inside it. If the container isn't running then `cat` command will fail making the test fail. 

I never considered the output of os-release to be that important in this scenario and I could have picked a different command. The reasoning for picking os-release is that it is present in every product and it's just:
```
NAME="SLES"
VERSION="15-SP4"
VERSION_ID="15.4"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP4"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp4"
DOCUMENTATION_URL="https://documentation.suse.com/"
```

This does not mean we shoulnd't follow your recommendation if it's important

grisu48 replied:
I see, then perhaps executing `true` instead of `cat` might be better comprehensible? But the current solution works, so if you like it, just ignore this comment :slightly_smiling_face:

grisu48 replied:
`cat /etc/os-release` suggests that someone would like to check if the values there make sense.

`true` suggests to me, this is just a smoke test.

m4u9 replied:
so just replacing `cat /etc/os-release` with `true` is enough?

grisu48 replied:
If it's just a smoke test, yes - unless you think that `cat /etc/os-release` is better suited!

m4u9 replied:
yea it's essentially a smoke test for now. The SUMA QE team may eventually decide to improve upon these tests but for now at maintenance we just want something simple to automate the container release.

m4u9 replied:
I replaced `cat /etc/os-release` with `true`

grisu48 reviewed (commented):
Some minor suggestions, overall it's getting there :slightly_smiling_face:

grisu48 reviewed (approved):
LGTM! Good job! :partying_face:

grisu48 commented:
We still need some verification runs, then it's ready to be merged.

m4u9 commented:
> We still need some verification runs, then it's ready to be merged.

What are the verification runs needed?

grisu48 commented:
> We still need some verification runs, then it's ready to be merged.

When we do changes to a openQA module or introduce a new one, we always want to have an openQA test run that shows, that everything is working.

jlausuch commented:
> > We still need some verification runs, then it's ready to be merged.
> 
> When we do changes to a openQA module or introduce a new one, we always want to have an openQA test run that shows, that everything is working.

He might need some help to do that... I don't think he has the right to trigger an isos post...

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,45 @@
+    elsif (check_var('CONTAINER_SUMA', 'chart')) {
+        script_retry("helm pull $image", timeout => 300, delay => 60, retry => 3);
+        assert_script_run("tar -tf proxy-*.tgz && rm proxy-*.tgz");
+    }
```
Let's catch configuration errors

```suggestion
    } else {
        die "Unsupported CONTAINER_SUMA setting";
    }
```

m4u9 replied:
done

grisu48 commented:
VRs:

* [suma-proxy-tftpd](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-tftpd&groupid=180)
* [suma-proxy-ssh](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-ssh&groupid=180)
* [suma-proxy-squid](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-squid&groupid=180)
* [suma-proxy-salt-broker](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-salt-broker&groupid=180)
* [suma-proxy-httpd](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-httpd&groupid=180)
* [suma-proxy-helm](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-helm&groupid=180)

They are still WIP.

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,45 @@
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
```
```suggestion
        my $runtime = get_required_var('CONTAINER_RUNTIME');
        script_retry("$runtime pull $image", timeout => 300, delay => 60, retry => 3);
```

Se need to use `CONTAINER_RUNTIME`, as we're also having `docker` test runs.

grisu48 replied:
See e.g. https://duck-norris.qe.suse.de/tests/13743#step/suma_containers/9 where on SLES12-SP5 the test run fails because we should be testing `docker` there and not `podman`.

m4u9 replied:
done

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,45 @@
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        script_retry("podman pull $image", timeout => 300, delay => 60, retry => 3);
+        assert_script_run("podman run --rm $image true");
```
```suggestion
        assert_script_run("$runtime run --rm $image true");
```

m4u9 replied:
done

grisu48 commented:
@m4u9 Can you please do a rebase and address the open issues? Most of the VR's look good so far, only some small things that are left.

m4u9 commented:
@grisu48 I rebased addressing the issues

grisu48 commented:
VRs restarted!

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,48 @@
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        my $runtime = get_required_var('CONTAINER_RUNTIME');
+        script_retry("$runtime pull $image", timeout => 300, delay => 60, retry => 3);
```
My previous remark that `$runtime` dan be a set of runtimes (e.g. `podman,docker`) somehow got lost and we need to re-implement this.

See e.g. https://duck-norris.qe.suse.de/tests/13835#step/suma_containers/9

grisu48 replied:
Weirdly this was working before, @m4u9 could it be that this somehow got overwritten? :thinking:

grisu48 replied:
We discussed this in Slack - Please just ping me if you still need help. We're almost there :slightly_smiling_face:

m4u9 commented:
> @m4u9 ping?

Sorry for the delay, I made some changes, let's see if it works

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,49 @@
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        my $runtime = get_required_var('CONTAINER_RUNTIME');
+        script_retry("for i in $(echo $runtime | tr ',' '\n'); do $i pull $image; done", timeout => 300, delay => 60, retry => 3);
```
```suggestion
        my @runtime = split /,/, $runtime;
        for $runtime (@runtime) {
            script_retry("$runtime pull $image", timeout => 300, delay => 60, retry => 3);
            assert_script_run("$runtime run --rm $image true");
        }
```

What do you think of this approach instead of using bash `for` loop?

grisu48 replied:
With this and a rebase I think we're done. If you agree, please update the PR and ping me for verification runs, then we can proceed and merge it.

m4u9 replied:
well I did it in bash because that's what I know, if you prefer to have it in Perl I have no objections, I just don't really know how to write Perl code ;)

grisu48 replied:
Understandable :slightly_smiling_face:

grisu48 commented:
VRs restarted:

* [suma-proxy-tftpd](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-tftpd&groupid=180)
* [suma-proxy-ssh](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-ssh&groupid=180)
* [suma-proxy-squid](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-squid&groupid=180)
* [suma-proxy-salt-broker](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-salt-broker&groupid=180)
* [suma-proxy-httpd](https://duck-norris.qe.suse.de/tests/overview?distri=sle&version=15-SP5&build=suma-proxy-httpd&groupid=180)
* [suma-proxy-helm](https://duck-norris.qe.suse.de/tests/overview?version=15-SP5&groupid=180&build=suma-proxy-helm&distri=sle)

grisu48 commented:
Waiting for rebase, then we can merge. The VRs look ok.

m4u9 commented:
> Waiting for rebase, then we can merge. The VRs look ok.

Rebase is done

grisu48 commented:
@m4u9 https://github.com/os-autoinst/os-autoinst-distri-opensuse/actions/runs/6603222696/job/17991336466?pr=17437 is failing. You need to run `tools/tidy`.

m4u9 commented:
> @m4u9 https://github.com/os-autoinst/os-autoinst-distri-opensuse/actions/runs/6603222696/job/17991336466?pr=17437 is failing. You need to run `tools/tidy`.
> 
> To avoid such issues, you can run `make test-static` on your local repo before pushing.

done, hopefully it will pass this time :)

grisu48 commented on tests/containers/suma_containers.pm:
```diff
@@ -0,0 +1,52 @@
+    my $image = get_required_var('CONTAINER_IMAGE_TO_TEST');
+
+    if (check_var('CONTAINER_SUMA', 'image')) {
+        my $runtime = get_required_var('CONTAINER_RUNTIME');
```
To solve the "Loop iterator is not lexical issue" we need to change the loop variable's scope to be only loop local. The easiest way is to rename this variable so that the loop variable `$runtime` is not declared outside.

```
        my $runtimes = get_required_var('CONTAINER_RUNTIME');
        my @runtimes = split /,/, $runtimes;
        for my $runtime (@runtimes) {
```

grisu48 commented:
VRs passing: https://duck-norris.qe.suse.de/tests/14275 https://duck-norris.qe.suse.de/tests/14276 https://duck-norris.qe.suse.de/tests/14277 https://duck-norris.qe.suse.de/tests/14278
